keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,576 @@
+  // This field is OPTIONAL.
+  string staging_target_path = 3;
+}
+```
+
+```
+message NodeGetVolumeStatsResponse {
+  // This field is OPTIONAL.
+  repeated VolumeUsage usage = 1;
+  // This field is OPTIONAL.
+  repeated VolumeHealth volume_health = 1;
+}
+```
+
+```
+message VolumeUsage {
+  enum Unit {
+    UNKNOWN = 0;
+    BYTES = 1;
+    INODES = 2;
+  }
+  // The available capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 available = 1;
+
+  // The total capacity in specified Unit. This field is REQUIRED.
+  // The value of this field MUST NOT be negative.
+  int64 total = 2;
+
+  // The used capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 used = 3;
+
+  // Units by which values are measured. This field is REQUIRED.
+  Unit unit = 4;
+}
+```
+
+```
+message VolumeHealth {
+  // The error code describing the health condition of the volume.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+Add a `GET_VOLUME_STATS_HEALTH` node capability. If driver supports this capability, it MUST fetch `volume_health` information in `NodeGetVolumeStats`.
+
+```
+message NodeGetCapabilitiesRequest {
+  // Intentionally empty.
+}
+```
+
+```
+message NodeGetCapabilitiesResponse {
+  // All the capabilities that the node service supports. This field
+  // is OPTIONAL.
+  repeated NodeServiceCapability capabilities = 1;
+}
+```
+
+```
+// Specifies a capability of the node service.
+message NodeServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+      STAGE_UNSTAGE_VOLUME = 1;
+      // If Plugin implements GET_VOLUME_STATS capability
+      // then it MUST implement NodeGetVolumeStats RPC
+      // call for fetching volume statistics.
+      GET_VOLUME_STATS = 2;
+      // See VolumeExpansion for details.
+      EXPAND_VOLUME = 3;
+      // If Plugin implements GET_VOLUME_STATS_HEALTH capability
+      // then it MUST implement NodeGetVolumeStats RPC
+      // call for fetching volume health information.
+      GET_VOLUME_STATS_HEALTH = 4;
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+### External controller
+
+#### CSI interface
+Call GetVolume() RPC for volumes periodically to check the health condition of volumes themselves.
+
+#### Node failure event
+Watch node failure events.
+In the case of a node failure, the controller will report an event for all PVCs on that node.
+For network storage in the case of a node failure, the controller will log an event.
+
+### External agent
+
+#### CSI interface
+Call NodeGetVolumeStats() RPC to check the mounting and other health conditions.
+
+Call both GetVolume() and NodeGetVolumeStats() RPCs for local storage when local storage CSI support is enabled.
```

Comment:
Or should local storage CSI driver also deploy the controller piece as a daemonset? Then we don't need to duplicate the responsibilities.

Reply:
The responsibilities should not be duplicate.  The monitoring controller will get information directly from the storage system so we don't need this type of information to be retrieved on every node. However, mount related information can only be retrieved on every node.

Reply:
By duplicate responsibilities I mean that 1) external controller is already calling GetVolume() and 2) node agent also calls GetVolume only in the local volume case.  Can we just have external controller call GetVolume() in all cases?

Reply:
Got it.  Will change.